### PR TITLE
Workaround redirect of index.html to /

### LINKED ISSUE #38: http://localhost:9100/index.html is redirecting to ./ - this break the auth flow
Screenshot from 2020-02-04 16-27-40

We should simply serve the index.html ...

### PR DESCRIPTION
**Pull request by DeepDiver1975**
Fixes #38

### CODE REVIEW

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/service/v0/service.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-phoenix/pullRequest?prid=4947915&bid=16354768)

File: pkg/service/v0/service.go [hunk 1]
```diff
@@ -2,15 +2,14 @@ package svc
...
-	"io/ioutil"
-	"net/http"
-	"os"
-	"strings"
-
 	"github.com/go-chi/chi"
 	"github.com/owncloud/ocis-phoenix/pkg/assets"
 	"github.com/owncloud/ocis-phoenix/pkg/config"
 	"github.com/owncloud/ocis-pkg/log"
+	"io/ioutil"
```
Comment (DeepDiver1975) [inline pkg/service/v0/service.go]: bloody ide ...

**Comment by ownclouders** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/service/v0/service.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/owncloud/ocis-phoenix/pullRequest?prid=4947915&bid=16354768)

**Comment by refs** [comment]
@DeepDiver1975 so I fired up my debugger because the redirect comes from [go's http package](https://golang.org/src/net/http/fs.go?s=19393:19451#L542):


```golang
   541  // name is '/'-separated, not filepath.Separator.
   542  func serveFile(w ResponseWriter, r *Request, fs FileSystem, name string, redirect bool) {
   543  	const indexPage = "/index.html"
   544  
   545  	// redirect .../index.html to .../
   546  	// can't use Redirect() because that would make the path absolute,
   547  	// which would be a problem running under StripPrefix
   548  	if strings.HasSuffix(r.URL.Path, indexPage) {
   549  		localRedirect(w, r, "./")
   550  		return
   551  	}
```

The root of the issue is where the OIDC callback points to:

```html
<!DOCTYPE html>
<html lang="en">
<script>
    window.onload = function() {
		window.location.href = window.location.origin +
			window.location.pathname.substr(0, window.location.pathname.length - 18) +
			"index.html#/oidc-callback" + window.location.search;
    }
</script>
</html>

```

it is the `index.html` that the go server will pick up and redirect to root, this is the root (pun intended) of the redirect.

We should instead update `oidc-callback.html` to:

```html
<!DOCTYPE html>
<html lang="en">
<script>
    window.onload = function() {
		window.location.href = window.location.origin +
			window.location.pathname.substr(0, window.location.pathname.length - 18) +
			"#/oidc-callback" + window.location.search;
    }
</script>
</html>

```

essentially removing the `index.html`, and that would work as intended.

cc @DeepDiver1975 

The rationale behind this is:

    ServeFile replies to the request with the contents of the named
    file or directory.
     
    If the provided file or directory name is a relative path, it is
    interpreted relative to the current directory and may ascend to
    parent directories. If the provided name is constructed from user
    input, it should be sanitized before calling ServeFile.
     
    As a precaution, ServeFile will reject requests where r.URL.Path
    contains a ".." path element; this protects against callers who
    might unsafely use filepath.Join on r.URL.Path without sanitizing
    it and then use that filepath.Join result as the name argument.
     
    As another special case, ServeFile redirects any request where r.URL.Path
    ends in "/index.html" to the same path, without the final
    "index.html". To avoid such redirects either modify the path or
    use ServeContent.
     
    Outside of those two special cases, ServeFile does not use
    r.URL.Path for selecting the file or directory to serve; only the
    file or directory provided in the name argument is used.

**Comment by DeepDiver1975** [comment]
> essentially removing the `index.html`, and that would work as intended.

No - because there are setups where / does not default to index.html. This will break there.

**Comment by refs** [comment]
@DeepDiver1975 not sure I get that. Is it a hard requirement that the OIDC url has to contain index.html on it’s path?

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 not sure I get that. Is it a hard requirement that the OIDC url has to contain index.html on it’s path?

All I'm saying is that using urls without index.html adds web server support to redirect from / to index.html. Go adds this in the other direction. Feels contradicting ...

**Comment by tboerger** [comment]
Web servers aren't redirecting from / to index.html at all.

**Comment by DeepDiver1975** [comment]
> Web servers aren't redirecting from / to index.html at all.

redirect was the wrong term. web servers serve index.html when request goes to /
some do some don't. as of now phoenix uses index.html explicitly to ensure proper operations without the requirement that index.html is served as /.

now clearer what I mean?

**Comment by tboerger** [comment]
Maybe we should just ignore the phoenix html file within the assets and serve that single html file on our own within ocis-phoenix.

**Pull request closed without merging**